### ISSUE TITLE
Group Consecutive, Non-Mention Notification Events

### ISSUE THREAD

**Issue by danirabbit** [mastodon/mastodon#1483]
## User research

I asked around a few friends about what they thought about grouped notifications, here are the responses:

> It’s been too long since I used the twitter implementation
>
> But I think I enjoyed that could tap on the user’s picture to go to their profile – I know what I wrote, I don’t need to see it every time
>
> I guess the way mastodon web does things is interesting too, where the post itself is embedded underneath the notification, and it’s a completely interactive post

> Well, let me first start by saying that I really really dislike the way Twitter groups them
>
> I have notifications on for important accounts so that I can skip checking my feed, and if more than one account posts between when I last checked, the new posts are grouped in one notification
>
> And it’s not clear to me how many of the notifications are new, since they’re just added to the row of profile photos. So I constantly miss notifications
>
> To be honest, I actually really like the way Ivory does it, I never miss anything. Only thing I don’t like is how mentions are in activity, I end up seeing all of those twice
>
> If I get a mention, it’ll be in both the mentions tab and activity tab
>
> I would most like Mentions & Everything except mentions
>
> Otherwise the everything tab is always frustrating because I typically check mentions first
>
> Also worth considering that I had totally different opinions on all this when I was a volume poster and had tons of engagement

## Design

[\_Notification-Type-Icon - Material Shared Components (Figma)](https://www.figma.com/file/SivYQCculwAs2kBTDUuqth/Material-Shared-Components?type=design&node-id=21-2207&mode=design&t=YIu73B2GMFqO7dXz-11)

(Android designs are left, **web is right**)

[\_Notification Header - Material Shared Components (Figma)](https://www.figma.com/file/SivYQCculwAs2kBTDUuqth/Material-Shared-Components?type=design&node-id=25-2693&mode=design&t=YIu73B2GMFqO7dXz-11)

[Notification Neue - Material Shared Components (Figma)](https://www.figma.com/file/SivYQCculwAs2kBTDUuqth/Material-Shared-Components?type=design&node-id=21-2221&mode=design&t=YIu73B2GMFqO7dXz-11)

[Notifications - Notifications (Figma)](https://www.figma.com/file/XL75cL60SbsaXaVfsl9yzU/Notifications?type=design&node-id=855-8053&mode=design&t=JtvtjgEulP8ZF7Ef-11)

## Acceptance Criteria

- [ ] The new notifications take on a similar layout to the post changes outlined in [AND-171](https://linear.app/mastodon/issue/AND-171/post-layout-changes)
  - [ ] Content is indented, an icon sits in the margin
- [ ] Unread notifications now have a `surface-container` background
  - [ ] Read is still `surface`
- [ ] Post actions are now color-coded, add these extended colors to the project:

  <img src="https://uploads.linear.app/fb74b89d-1119-45e4-8dfe-3f87b2d742ec/5f528b8a-bc12-4e76-924c-7353e4867e1e/dfbaf228-da6b-45a1-b718-a40960188529?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2ZiNzRiODlkLTExMTktNDVlNC04ZGZlLTNmODdiMmQ3NDJlYy81ZjUyOGI4YS1iYzEyLTRlNzYtOTI0Yy03MzUzZTQ4NjdlMWUvZGZiYWYyMjgtZGE2Yi00NWExLWI3MTgtYTQwOTYwMTg4NTI5IiwiaWF0IjoxNzE1MTc3NTEzLCJleHAiOjMzMjg1NzM3NTEzfQ._zxGmrbXzhIPZbVFLV-518nlIeq--eflrlZTtou2uPM " alt="Screenshot 2024-05-07 at 2.29.58 PM.png" width="1025" height="546" />
- [ ] For each notification cell, the container for the leading icon (the one in the margins) takes up the same amount of horizontal space as the avatar so they line up.
  - [ ] Inside the container, the icon is 24pt large, filled, and centered within its container
  - [ ] In the designs, I'm showing a new post cell style. This is not strictly necessary for this feature to ship, but I just request that the icon container remains aligned with the avatar container. See this x-ray ruler line diagram:

    <img src="https://uploads.linear.app/fb74b89d-1119-45e4-8dfe-3f87b2d742ec/112e6dc3-12cc-427a-b636-37a96e9d375a/bb59098a-7f23-450c-83ab-bedf07114e08?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2ZiNzRiODlkLTExMTktNDVlNC04ZGZlLTNmODdiMmQ3NDJlYy8xMTJlNmRjMy0xMmNjLTQyN2EtYjYzNi0zN2E5NmU5ZDM3NWEvYmI1OTA5OGEtN2YyMy00NTBjLTgzYWItYmVkZjA3MTE0ZTA4IiwiaWF0IjoxNzE1MTc3NTEzLCJleHAiOjMzMjg1NzM3NTEzfQ.wuNkz1D_mc70kH4JfbGUX1bCfF-mHZCeSHqM6Lwrns4 " alt="image.png" width="1984" height="2120" />
- [ ] Notification header text ("Firstname followed you") is Material body medium
  - [ ] Names and server domains are bolded
  - [ ] Whenever there is a headline/body style lockup (i.e. server announcements), the headline uses Material title small.
- [ ] Avatars and bolded names are links to their respective profiles
- [ ] Clicking anywhere else in the cell links to the post, where that post uses the highlighted presentation.
  - [ ] **The exception** is the grouped follower notification, whose cell links to your followers list.

**Comment by chris-martin** [comment]
This presents a dangerous temptation to start showing events out of order like Twitter does.

**Comment by danrabbit** [comment]
@chris-martin I don't think these things are related. I would not like to see things out of order.

**Comment by pawelngei** [comment]
1 . I just got 32 notifications of people faving / boosting the same toot. I think that those things could be merged into two messages: 32 people boosting, 32 people faving, and the chronology of those is irrelevant.

2. I think we should also allow people to group the boosting / faving different toots in some discrete periods set in the Preferences. Let's say 0, 10, 15, 30, 60, 120 and 180 minutes.

This way if my lasts three toots become really popular in the last 2 hours, I would see 6 notifications instead of dozens.

No Twitter-like manipulation, just grouping as set by the user. Changes purely in the display / client side.

**Comment by rysiekpl** [comment]
I love @pawelngei 's take on this. This would be very, very useful.

**Comment by nightpool** [comment]
"brand account"

**Comment by pawelngei** [comment]
@nightpool whatever the original intention, this feature would be really, really useful.

Please let's not go the Diaspora* way, where people were really hostile to anybody who wanted to use their software for anything but Free Software Evangelism™. Building a network for people means adding a lot of functions useful for different kinds of people, some of whom we don't agree with.

**Comment by nightpool** [comment]
@pawelngei there's a wide gulf of middle ground between "making fun of corporations trying to co-opt a community to sell their products" and "being hostile to anybody who want[s] to use [our] software for anything but Free Software Evangelism", and i can support one while condemning the other.

**Comment by rysiekpl** [comment]
> and i can support one while condemning the other.

Then please start doing so. So far your contribution to this discussion seems to be shooting down a good idea just because it might be useful in a particular use-case you don't seem to like much.

I would love this functionality on my *personal account*. Brand thing is completely irrelevant here. Can we please focus on the merits of the proposal, instead of knee-jerking like this? It doesn't help any.

**Comment by tykayn** [comment]
it could be done by only changing the frontend, it should be enough to group things in a very pleasant and fast delivery at first

**Comment by pawelngei** [comment]
> it could be done by only changing the frontend, it should be enough to group things in a very pleasant and fast delivery at first

One, don't give me ideas, I have too many projects in parallel and I might just do it!

Two, wouldn't it be cleaner if the backend grouped it based on user's input?

**Comment by tykayn** [comment]
haha, let the dark side of the project give you side projects
i guess it would be cleaner, but would need to touch way more things and maybe break more things if the object sent by the api changes.
i don't know the backend well at all.

but on the frontend it would only change the display of the list of notifications. :)

**Comment by tykayn** [comment]
there is no in house solution for this in the web app of mastodon.
i developed something to achieve kind of this multiple account posting here
https://forge.chapril.org/tykayn/multi-account-post-schedule-mastodon

but it requires to make a developer api token for each account you want to use, save them in an environment file .env like said in the README.md ,
run this node project, and tada you get multiple account webpage.
you can switch from one account to an other with a selector.

bonus: there is a selector for schedule posting date. (i guess it would also be great to have the possibility to schedule toots in the web app of mastodon, it looks like something very easy to add. only difficulty i see would be UX to not clogger the view with all the options displayed like we are in Final Form Excel)

far from ideal, but it works and could inspire some merge request :D

**Comment by yuvalne** [comment]
@pawelngei from what I've seen, the consensus is simply that they should be grouped. I agree that maybe eternal backwards grouping should not be on by default (although idk, when I see a single notification saying "53 people liked your toot" a year after is was tooted I can deduce that it came later than the previous ones), but I think your suggestion in [this comment](https://github.com/mastodon/mastodon/issues/1483#issuecomment-436656623) will probably be best: allow grouping of notifications within the past x minutes, with the default being 180 and the options being, let's say, 0, 5, 10, 30, 60, 120, 180 and infinity.
The specific numbers don't matter all that much, I think any selection with a reasonable amount of options will be good.

**Comment by DavidBuchanan314** [comment]
I just thought I'd poke my head in and say that this feature would be increasingly useful, as the network grows.

Once a post has more than a few hundred likes, and they're coming in at a rate of several-per-minute, I wouldn't mind waiting N minutes (for some value of N around 10-15) for a batch of notifications to be withheld and then delivered as one. This would increase the latency of notifications - but does waiting an extra 10 minutes to receive your 1001th like really matter?

Alternatively, I wouldn't mind if the N most recent (or within some timeframe) notifications of the same type were coalesced retroactively, entirely in the frontend. This might give the impression of things being reordered, but I think that's a minor inconvenience. To illustrate what I mean by reordering, if my notifications looked like this:

- userA and 37 others liked post X
- userB and 23 others repeated post X

And then, after userC repeats post X, they'll now look like:

- userC and 24 others repeated post X
- userA and 37 others liked post X

Whatever approach is taken, ideally the various thresholds should be configurable, to suit what kind of audience and/or notification-spam-tolerance a given user has.

(disclaimer, I'm actually a pleroma user, but I believe everything I've said here applies equally to mastodon)

**Comment by nehemiagurl** [comment]
I agree with @DavidBuchanan314 that the reordering would be a minor inconvenience, whether it is grouped by time or by number. The real question would be how hard it would be to implement this.

**Comment by eastein** [comment]
This would be really useful. I barely use Mastodon (just joined) and the notifications are already a bit over the top. They'd be much more manageable grouped.

**Comment by lennym** [comment]
Just wanted to add my support for this. I recently had a post which received several hundred boosts/favourites, and it made the notifications column basically unusable.

Being able to filter for mentions is useful, but not being able to see notifications for new followers is really unfortunate, so maybe if those were able to be filtered as well then that might solve some of the problem.

**Comment by andreykurenkov** [comment]
Also want to add support. Some people feel very strongly about this and think Mastodon is not pleasant to use because this feature is not present: https://sigmoid.social/@lb/109407086341764057

If the concern is focus on virality/numbers, another option could be added to not show the exact number (eg "userA and many others")  - some Mastodon apps already have an option for the user to disable the display of number of favorites/boosts on a given post, and this is in the same spirit.

**Comment by timdmackey** [comment]
I made the mistake (lol) of posting a thoughtful reply yesterday on a post from a relatively popular account. Ever since then, my notifications have been pretty much unusable in the apps I use. In the web app I was able to turn off boost and favorite notifications, so that has definitely helped.

Regarding other apps, this being an issue does point to the possibility mentioned above of this being solved at the API level rather than just in the UI.

**Comment by vibeisveryo** [comment]
From my post at #22578 

### Pitch

Instead of, as is:

Notifications
- Alex favourited your post (@vibeisveryo: Hello!)
- Bella favourited your post (@vibeisveryo: Hello!)
- Bella boosted your post (@vibeisveryo: Hello!)
- Bella favourited your post (@vibeisveryo: Goodbye!)
- Carmen favourited your post (@vibeisveryo: Hello!)
- David favourited your post (@vibeisveryo: Hello!)
- David favourited your post (@vibeisveryo: Goodbye!)
- Emmanuel favourited your post (@vibeisveryo: Hello!)

Group the same action which does not carry information other than the actor (i.e. not replies, but boosts and favourites) in the same time period (say, 24h, maybe configurable) together. I do not currently see an option for this, and have 10 notifications I have to scroll through which are mostly redundant. Post change:

Notifications
- 5 people favourited your post (@vibeisveryo: Hello!) *Note: hovering over 5 people would show their names, and the UI element would probably show their avatars*
- Bella boosted your post (@vibeisveryo: Hello!)
- 2 people favourited your post (@vibeisveryo: Goodbye!)

Alternatively for the first one, say something like Alex, Bella, and 3 others favourited your post.

### Motivation

Reduce visual clutter in notifications and make it easier to separate notifications needing or inviting further engagement (replies) from notifications not doing so (favourites)

**Comment by albjeremias** [comment]
something like thissss!!!
![image](https://user-images.githubusercontent.com/26681446/209604123-c372baf8-e6c3-4d12-baf7-f852b682bb47.png)

also change the private messages to something like:
![image](https://user-images.githubusercontent.com/26681446/209606433-0954541c-0efb-41ad-bcd7-bd631edb817e.png)
and only show the content of the message when clicked, or when the dm "tab" is open?!

[design on penpot](https://design.penpot.app/#/view/9c5ceddf-c68c-8036-8001-cf95c95013ce?page-id=9c5ceddf-c68c-8036-8001-cf95c95013cf&section=interactions&index=0&share-id=35dd4c70-bf14-80a7-8001-cfa502432c90)

design using https://penpot.app

**Comment by paulgp** [comment]
Worth noting that [Fedilab (on Android)](https://play.google.com/store/apps/details?id=app.fedilab.android&hl=en_US&gl=US&pli=1) has adopted some verison of this approach with its notifications:

![image](https://user-images.githubusercontent.com/4665750/210396612-7aa8b11d-9e67-4cd4-a919-a56bcf218741.png)

**Comment by Justinzobel** [comment]
Now that Mastodon has come into it's stride I'd love to see this implemented. It's difficult to see what is happening when you get 60 or 70 notifications and have to scroll through them to see what each is.

**Comment by elysdir** [comment]
I too would love to see this implemented. To me, it doesn’t matter much _how_ it’s implemented (all of the proposals I’ve seen here look fine to me); I just want my notifications list (especially in the web front end) to be more usable.

My notifications over the past two days consist of 17 favourites, 13 boosts, and 2 other items. It’s lovely to see people favouriting and boosting my posts, but I don’t need an individual notification for each one of those; it’s hard to find the more important things among the favourites and boosts. And I can only imagine that for really popular posts, notifications must become entirely unusable.

**Comment by LoVicente** [comment]
This would be very useful! since the alternative is already implemented, it could be an optional setting on the client. Assuming there are people who would prefer to see the individual notifications.

**Comment by Pxtl** [comment]
+1 - even under 100 likes/retoots can completely destroy your notifications.

**Comment by albjeremias** [comment]
> I am getting new notifications for a particular viral post every minute or so, sometimes every few seconds. Mastodon needs a feature to deal with this problem.
> 
> Not seeing boosts or favorites is not an answer, because this is not the only post I have made that is getting boosts and faves, and I still want to see those.
> 
> Edit: I got 9 more while I was typing that in.

u can use fedilab.. as mentioned in https://github.com/mastodon/mastodon/issues/1483#issuecomment-1369957020

**Comment by DavidBuchanan314** [comment]
By the way, bsky.app (the official bluesky client, MIT licensed) does notification grouping entirely within the client front-end, and it works pretty well - I'm not sure it necessarily needs work in every layer of the stack, although it would perhaps work better if it did.

https://github.com/bluesky-social/social-app/blob/9750c393a00ad935fe0e7278572e693016ba7e30/src/state/models/feeds/notifications.ts#L644-L672

**Comment by whitequark** [comment]
> As discussed above, it needs to be implemented on the backend, in Rails, touching a lot of parts of Mastodon, so it's unlikely it can be a simple PR that @Gargron could just merge in.

This *was* implemented, wasn't it? https://github.com/mastodon/mastodon/pull/11446

**Comment by pawelngei** [comment]
> This _was_ implemented, wasn't it? #11446

If I understand the thread correctly, it was not merged, it was closed. 

Even it if would be merged, what was implemented here was JS-only, as I outlined above. The changes I'm talking about need to happen in the backend, Rails part of Mastodon, are much trickier and more complex. As you can see in the thread, Gargron is not a fan:

> We could just not merge it and close the issues asking for it. I don't mind individual notifications.

https://github.com/mastodon/mastodon/pull/11446#issuecomment-516297592

**Comment by whitequark** [comment]
Sure. I didn't say it was merged. I'm saying that it was implemented without needing the Rails backend modifications that are contentious. So why not merge a JS-only variant?

I have almost 6K followers (down from 26K I used to have elsewhere) and the notifications are... tricky. I do in fact keep up with all of them individually; the problems begin when I'm trying to *navigate* the stream of notifications to find a specific one.

**Comment by pawelngei** [comment]
I'm afraid that all I can say is "because the maintainers don't want to have a discussion on the possible design of this functionality".

If I was to implement it myself, I would probably do it in one of the forks of Mastodon and if it gets merged, show it as a proof-of-concept which could be possibly ported into the Mastodon itself. Maybe https://github.com/hometown-fork/hometown or similar?

Sadly, no time for FLOSS contributions in the next few weeks / months.

**Comment by ronilaukkarinen** [comment]
+1 to this. This is very needed. What amazes me is the fact that this was already implemented, but never merged as stated [here](https://github.com/mastodon/mastodon-android/issues/160#issuecomment-1127081759).

> I'm not sure this has to happen on the API side. I implemented it for the web app in https://github.com/mastodon/mastodon/pull/11446 but gave up on it due to the nitpicking in the responses.

I don't know why anyone would want this to be opt-in. This should be default.

**Comment by brendanjones** [comment]
This is obviously achievable purely in the client, given multiple apps support this. I don’t know how anyone uses notifications without grouping! As an example, screenshots of Mona and Woolly, in that order. Notice how Mona groups all notifications from one post, while Woolly breaks them up after a certain time.

![IMG_7180](https://github.com/mastodon/mastodon/assets/16049594/35a68db2-5b97-4d97-a2e9-835878e2922a)
![IMG_7181](https://github.com/mastodon/mastodon/assets/16049594/f98d1891-4ecc-49a4-a661-3c6d9a44cc4c)

These are just two examples, I’m sure there are more. Both are fine as solutions, I don’t think we need to quibble much over the exact grouping method. If breaking them up by time I’d say do it once a day, but anything from 1hr-24hrs is fine honestly.

**Comment by delusho** [comment]
I would love to have this and it would make the mastodon app so much more usable.

**Comment by Sigma-One** [comment]
especially recently i've had some toots gain a good bit of traction, which makes the notification column just useless for a while, so :+1: from me to this feature

**Comment by ronilaukkarinen** [comment]
Re-bumping this. I've come to enjoy this grouping when using Mona on iPhone. Attempted to implement it myself to web UI but I'm not skilled enough. One of my posts went viral yesterday and getting hundreds of notifications, it would be fun to see "x users favorited your posts" instead of getting them individually. You can always disable notifications, but then what's the point? Any JS gurus out there to send another PR based on [this one](https://github.com/mastodon/mastodon/pull/11446/commits/7e31cce96db080bb26200021d7a5679469335819). Tested that one out as well but I guess it's so old it won't do anything without modifications in the latest version.

**Comment by Pxtl** [comment]
I've read a couple of "internet famous" people griping about this.  Their notification tabs become unusable and so they have to fuss with the filters.  It's going to chase them off of the platform.

**Comment by resuna** [comment]
And my notifications blew up again. AUGH. This ticket was opened over six years ago, man, just do it.

**Comment by thanius** [comment]
This can and should be implemented as an option on frontend.

**Comment by thanius** [comment]
Above examples should suffice, however here's my take on it.

Replies/mentions are communication, so those ought to be separated from this suggestion.

Favorites and boosts though are totally irrelevant as **_when_** they actually arrived because there's no need of any action from OP, thus chronological order on these are not as important in the aggregated view. 

I suggest the following aggregated text on notifications:

**QUANTITIES TURNED ON**
**"User X and N others favorited your post"**
Clicking **User X** would take you to the profile as today, while clicking **N others** would take you to a new view which lists all the notifications, grouped by the post, in their chronological order as it's being presented today.

**QUANTITIES TURNED OFF**
**"User X and others favorited your post"**
Here we would same as before click on **others** to reach the full list of notifications for the post.

Any thumbnails should be cropped as to take less space in the notification list.

So to recap:
* In the backend we have a tuple containing the data - postid, latest interaction, aggregated values as of current timestamp/latest load as well as status of the notification (read/unread). And of course whatever else needed for notifications to work
* Chronological order on notifications of favorites/boosts are not relevant in an aggregated view, only the latest (reverse chronological order) or first (chronological order) person would be shown here
* All subsequent favorites/boosts would be appended to the string, with a link to the full list
* Thumbnails are cropped

If this were to be presented to the user as an opt-in option there would be no change for those who wants the notifications as today.

**Comment by pawelngei** [comment]
As @renchap pointed out [in this post](https://oisaur.com/@renchap/111516704367951336), this should happen in 4.3:

> We are working on improving notifications in 4.3, and this includes notifications grouping. This will most likely be done server-side.

**Comment by brendanjones** [comment]
> As @renchap pointed out [in this post](https://oisaur.com/@renchap/111516704367951336), this should happen in 4.3:
> 
> > We are working on improving notifications in 4.3, and this includes notifications grouping. This will most likely be done server-side.

Can this then be added to https://github.com/mastodon/mastodon/milestone/4 @renchap ?

**Comment by nehemiagurl** [comment]
[The latest blogpost](https://blog.joinmastodon.org/2024/05/trunk-tidbits-april-2024/) suggests this will be part of the 4.3 beta!

**Comment by whitequark** [comment]
:tada: Thank you for working on this!
